src/Features/Core/Portable/SignatureHelp/SignatureHelpServiceWithProviders.cs

```diff
@@ -0,0 +1,199 @@
+        {
+            if (_testProviders.Length > 0)
+            {
+                return _testAugmentBuiltInProviders
+                    ? GetBuiltInProviders().Concat(_testProviders)
+                    : _testProviders;
+            }
+            else
+            {
+                return GetBuiltInProviders();
+            }
+        }
+
+        public override bool ShouldTriggerSignatureHelp(
+            SourceText text,
+            int caretPosition,
+            SignatureHelpTrigger trigger,
+            OptionSet options = null)
+        {
+            if (trigger.Kind == SignatureHelpTriggerKind.Insertion)
+            {
+                foreach (var provider in GetProviders())
+                {
+                    if (provider.IsTriggerCharacter(trigger.Character))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private void FilterTextuallyTriggeredProviders(
+            ImmutableArray<SignatureHelpProvider> providers,
+            char ch,
+            out ImmutableArray<SignatureHelpProvider> triggeredProviders,
+            out ImmutableArray<SignatureHelpProvider> untriggeredProviders)
+        {
+            var triggeredBuilder = ImmutableArray.CreateBuilder<SignatureHelpProvider>();
+            var untriggeredBuilder = ImmutableArray.CreateBuilder<SignatureHelpProvider>();
+
+            foreach (var provider in providers)
+            {
+                if (provider.IsTriggerCharacter(ch))
+                {
+                    triggeredBuilder.Add(provider);
+                }
+                else
+                {
+                    untriggeredBuilder.Add(provider);
+                }
+            }
+
+            triggeredProviders = triggeredBuilder.ToImmutable();
+            untriggeredProviders = untriggeredBuilder.ToImmutable();
+        }
+
+        public override async Task<SignatureList> GetSignaturesAsync(
+            Document document,
+            int caretPosition,
+            SignatureHelpTrigger trigger = default(SignatureHelpTrigger),
+            OptionSet options = null,
+            CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var allProviders = GetProviders();
+            if (allProviders.IsEmpty)
+            {
+                return null;
+            }
+
+            switch (trigger.Kind)
+            {
+                case SignatureHelpTriggerKind.Insertion:
+                    // TODO(DustinCa): Clean up comment
+
+                    // Separate the sig help providers into two buckets; one bucket for those that were triggered
+                    // by the typed character, and those that weren't.  To keep our queries to a minimum, we first
+                    // check with the textually triggered providers.  If none of those produced any sig help items
```

Comment:
>If none of those produced any sig help items

If none of hte triggered providers produced anything, then i would think we would not have sig help at all.